improving is_perfect and is_mersenne_prime methods

mohitacecode opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #16144 
#### Brief description of what is fixed or changed
Perfomance of `is_perfect` and `is_mersenne_prime` is improved as suggested in pr(https://github.com/sympy/sympy/pull/16145#issuecomment-471144618)

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

Commit:
suggested changes

ethankward commented on sympy/ntheory/factor_.py:
```diff
@@ -28,6 +28,17 @@
+    global PERFECT
+    if PERFECT is None:
+        PERFECT = [2**(i - 1)*(2**i - 1) for i in MERSENNE_PRIME_EXPONENTS]
+    return n in PERFECT
```
This takes a large amount of memory- these are massive numbers. I think you should find a better way to cache this.

mohitacecode replied:
It was discussed that we should do whatever faster is here. (https://github.com/sympy/sympy/pull/16145#issuecomment-470649409).

asmeurer replied:
We could recompute them each time as well. It depends on whether we consider this to be too much memory usage.

mohitacecode replied:
If we are in favour of saving memory then we can use something like `python generators` and can generate the term whenever we required.(Will save lot of memory)
But all this process will take too much time and IMO we should give priority to time as no. of terms are limited here.

smichr replied:
An even number is perfect if it can be expressed as two factors: one a power of two and the other being one less than a related power of two. Except for the first 23 Mersenne prime powers it would be much faster to check the exponent:
```python
def _isperfect(n):
    if n%2 == 0:
        i = multiplicity(2, n)
        o = n >> i
        return multiplicity(2, o + 1) == i + 1
```
So I would suggest precalculating the first 23 perfect numbers and then for a given number do
```python
PERFECT = [2**(i-1)*(2**i-1) for i in MERSENNE_PRIME_EXPONENTS[:23]]
def _isperfect(n):
    global perfect
    if n % 2 == 0:
        while n > PERFECT[-1] and len(PERFECT) < len(MERSENNE_PRIME_EXPONENTS):
            i = 2**(MERSENNE_PRIME_EXPONENTS[len(PERFECT)] - 1)
            PERFECT.append(i*(2*i - 1))
        if n <= PERFECT[-1]:
            return n in PERFECT
    else:
        i,_ = integer_log(n, 2)
        if i < exponent below which all odds are known to be not perfect:
            return False
    # handle new analysis for odd or even
```

ethankward replied:
It takes 5 seconds to initialize `PERFECT = [2**(i - 1)*(2**i - 1) for i in MERSENNE_PRIME_EXPONENTS]` on a powerful computer.I don't know what you mean by "priority to time" but that is not fast.

asmeurer replied:
I think that's the point. If we cache the values that 5 seconds only needs to be run once. But it requires storing the numbers in memory. 

Personally I think that

- the 100 or so MB that this takes isn't so bad on a modern computer. But neither is 5 seconds.
- very few people will actually call these functions. I don't think they would be used in other places in SymPy (or am I wrong?). So it doesn't actually matter that much.

ethankward replied:
With the incremental caching it's much better.

mohitacecode commented on sympy/ntheory/factor_.py:
```diff
@@ -2222,30 +2233,90 @@ def is_perfect(n):
+            return False
+        if c <= 100:
+            return False
+        if a > integer_root(3*n, 3):
```
does it mean `integer_nthroot` because we dont have any method called `integer_root`.

mohitacecode replied:
what about this. @asmeurer

smichr replied:
I rewrote this and committed.

Commit:
fixes errors

smichr commented on sympy/ntheory/factor_.py:
```diff
@@ -2222,30 +2234,90 @@ def is_perfect(n):
+        if not any(n%m == r for m, r in [(12, 1), (468, 117), (324, 81)]):
+            return False
+        facs = factorint(n)
+        odd = [i for i in facs if i%2]
```
```suggestion
        odd = [(b, e) for b, e in facs.items() if e%2]  # only 1 odd exponent
```

smichr committed:
Update factor_.py

smichr committed:
compute Mersenne and perfect numbers as needed

This is more lazy than before because now they are only extended when a value
of n is greater than the current value in the lists.

mohitacecode commented:
@smichr Have you done all the changes? sorry I was working on different pr.

smichr committed:
keep only simplest tests for perfect odd numbers

These are conditions the factors must meet:
```
        facs = factorint(n)
      # length conditions
        if len(facs) < 10:
            return False
        if 3 not in facs and len(facs) < 12:
            return False
      # n = q**a*prod(p[i]**(2*e[i]), (i, 1, k)) where a is odd
        odd = [i for i in facs.items() if i[1]%2]
        if len(odd) != 1:
            return False
        q, a = odd[0]
        if a % 4 != 1 or q % 4 != 1:
            return False
        if set(facs.values()) == {a, 2}:
            return False  # all the even exponents cannot be 2
        del q, a, odd
      # criteria for small and large values
        k = len(facs) - 1
        _sorted = sorted(facs)
        if 3*_sorted[0] >= 2*k + 8:
            return False
        esum = sum(facs.values())
        if esum < 101 or esum*8 < 21*k - 18:
            return False
        c, b, a = _sorted[-3:]
        if c <= 100:
            return False
        if not (10**20 < b**5 < 2*n):
            return False
        if not (10**24 < a**3 < 3*n):
            return False
      # criteria for divisors with single prime factor
        if not any(b**e > 10**62 for b, e in facs.items()):
            return False
      # criteria for number of factors with even exponents
        k2 = 2**(k + 1)
        if integer_log(n, 2)[0] > k2*(k2 - 1):  # n <= 2**(4**(k+1) - 2**(k+1))
            return False
      # criteria based on product of exponents
        if prod(facs.values()) >= (2*n**Rational(17, 26)):
            return False
```

smichr committed:
new upper limit for OPN

smichr commented:
> @smichr Have you done all the changes? sorry I was working on different pr.

yes, see what you think.

smichr committed:
Update factor_.py

smichr committed:
work with int k in divisor_sigma

factorint returns ints to work with; native Python ints will be faster to compute with
than Integers

smichr committed:
update imports

smichr committed:
keep non-integer k handling

smichr commented:
This needs a final ok from @ethankward

smichr reviewed (approved):
I see that @ethankward already marked the changes as acceptable so I think this is ready to go. Will wait for any other feedback, though.

smichr commented:
Nope...looks like once someone has reviewed and requested changes they have to sign off on them. So this is waiting for @ethankward
